File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
import math
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        
        n=len(nums)
        result = [1]*n
        left_prod =1
        for i in range(n):
            result[i]=left_prod
            left_prod*=nums[i]
        right_prod=1
        for i in range(n):
            result[i]*=right_prod
            right_prod*= nums[i]
            

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.ts
function productExceptSelf(nums: number[]): number[] {
    let n = nums.length;
    const result = new Array(n);
    // Calculate prefix products
    let prefixProduct = 1;
    for (let i =0;i<n;i++){
        result[i]=prefixProduct;
        prefixProduct *=nums[i];

    }
    // Calculate suffix products and finalize the result 
    let suffixProduct =1;
    for (let i=n-1;i>=0;i--){
        result[i]*=suffixProduct;
        suffixProduct *=nums[i];
    }
    return result;
};